### PR TITLE
[DeviceSanitizer] Refactor the code to manage shadow memory

### PR DESCRIPTION
**Pull request by zhaomaosu**
* Added a ShadowMemory class to manage all related operations
* Release physical memory page if there is no dependencies

### CODE REVIEW

File: source/loader/layers/sanitizer/asan_shadow.cpp [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+                    getContext()->logger.error("urPhysicalMemCreate(): {}",
+                                               URes);
+                    return URes;
+                }
+
+                URes = getContext()->urDdiTable.VirtualMem.pfnMap(
+                    Context, (void *)MappedPtr, PageSize, PhysicalMem, 0,
+                    UR_VIRTUAL_MEM_ACCESS_FLAG_READ_WRITE);
+                if (URes != UR_RESULT_SUCCESS) {
+                    getContext()->logger.debug("urVirtualMemMap({}, {}): {}",
```
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/asan_shadow.cpp:138]:
```suggestion
                    getContext()->logger.error("urVirtualMemMap({}, {}): {}",
```

File: source/loader/layers/sanitizer/asan_shadow.cpp [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+                if (URes != UR_RESULT_SUCCESS) {
+                    getContext()->logger.error("EnqueueUSMBlockingSet(): {}",
+                                               URes);
+                    return URes;
+                }
+
+                VirtualMemMaps[MappedPtr].first = PhysicalMem;
+            }
+
+            // We don't need to record virtual memory map for null pointer.
```
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/asan_shadow.cpp:160]: Why not record this?

File: source/loader/layers/sanitizer/asan_shadow.cpp (hunk 1 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_shadow.cpp]: Done

File: source/loader/layers/sanitizer/asan_shadow.cpp (hunk 2 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_shadow.cpp]: for null pointer, it doesn't have an alloc info.

Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_shadow.cpp]: I also feel a little confused while reading this comment. 
I think we need to use nullpointer's shadow memory.

File: source/loader/layers/sanitizer/asan_shadow.cpp [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        ManagedQueue Queue(Context, Device);
+
+        auto URes = EnqueuePoisonShadow(Queue, 0, 1, kNullPointerRedzoneMagic);
+        if (URes != UR_RESULT_SUCCESS) {
+            getContext()->logger.error("EnqueuePoisonShadow(NullPointerRZ): {}",
+                                       URes);
+            return URes;
+        }
+        return URes;
+        return Result;
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_shadow.cpp:113]:
```suggestion
        return URes;
```

File: source/loader/layers/sanitizer/asan_shadow.cpp [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    }();
+    assert(Begin != 0 && End != 0);
+    ShadowBegin = Begin;
+    ShadowEnd = End;
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_shadow.cpp:117]:
```suggestion
```

File: source/loader/layers/sanitizer/asan_shadow.cpp (hunk 3 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_shadow.cpp]: Done

File: source/loader/layers/sanitizer/asan_shadow.cpp (hunk 4 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_shadow.cpp]: removed

File: source/loader/layers/sanitizer/asan_shadow.cpp (hunk 2 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_shadow.cpp]: Added more comments to describe the reasons.

**Review by AllanZyne — APPROVED** [review]

File: source/loader/layers/sanitizer/asan_shadow.cpp [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        static const size_t PageSize =
+            GetVirtualMemGranularity(Context, Device);
+
+        ur_physical_mem_properties_t Desc{
+            UR_STRUCTURE_TYPE_PHYSICAL_MEM_PROPERTIES, nullptr, 0};
+
+        // Make sure [Ptr, Ptr + Size] is mapped to physical memory
+        for (auto MappedPtr = RoundDownTo(ShadowBegin, PageSize);
+             MappedPtr <= ShadowEnd; MappedPtr += PageSize) {
+            std::scoped_lock<ur_mutex> Guard(VirtualMemMapsMutex);
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_shadow.cpp:160]: IMO, the lock range is too large here, it'd be better to lock the operations used on VirtualMemMaps only.

File: source/loader/layers/sanitizer/asan_shadow.cpp [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+            // We don't need to record virtual memory map for null pointer,
+            // since it doesn't have an alloc info.
+            if (Ptr == 0) {
+                continue;
+            }
+
+            auto AllocInfoIt =
+                getContext()->interceptor->findAllocInfoByAddress(Ptr);
+            assert(AllocInfoIt);
+            VirtualMemMaps[MappedPtr].second.insert((*AllocInfoIt)->second);
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_shadow.cpp:206]: We may also need an lock for unordered_set

**Review by yingcong-wu — APPROVED** [review]

**Comment by zhaomaosu** [comment]
Hi @oneapi-src/unified-runtime-maintain, could you please help review this PR? Thanks.

**Comment by zhaomaosu** [comment]
> please rebase

Done. The pre-ci failures are unrelated.

**Comment by zhaomaosu** [comment]
Hi @pbalcer, could you please raise the priority of merging this PR? We have many other works that will depend on this change. Many thanks.

**Comment by pbalcer** [comment]
> Hi @pbalcer, could you please raise the priority of merging this PR? We have many other works that will depend on this change. Many thanks.

done, can you please create an intel/llvm PR that updates UR to latest version?

**Comment by zhaomaosu** [comment]
> done, can you please create an intel/llvm PR that updates UR to latest version?

sure, PR created https://github.com/intel/llvm/pull/15736

**Pull request merged**